File: Archive#00-001-100/061-Rotate List/leetcode061.cpp
// <!-- encoding UTF-8 --!>
/*****************************************************************************
*                     ----Stay Hungry Stay Foolish----                       *
*   @author :   Shen                                                         *
*   @name   :   Leetcode 061                                                 *
*****************************************************************************/

#include <bits/stdc++.h>
using namespace std;
typedef long long int64;
template<class T>inline bool updateMin(T& a, T b){ return a > b ? a = b, 1: 0; }
template<class T>inline bool updateMax(T& a, T b){ return a < b ? a = b, 1: 0; }

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *   int val;
 *   ListNode *next;
 *   ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution
{
public:
    ListNode *rotateRight(ListNode *head, int k)
    {
        ListNode *cur = head;
        int cnt = 0, len = 0;
        while (cur != NULL)
        {
            if (cur->next == NULL)
            {
                len = ++cnt;
                cur->next = head;
                cur = head;
                cnt = 0;
                k %= len;
            }
            if (cnt == len - k - 1)
            {
                ListNode *ret = cur->next;
                cur->next = nullptr;
                return ret;
            }
            cnt++;
            cur = cur->next;
        }
        return head;
    }
};

void test() {}

int main() { test(); return 0; }

File: Archive#00-001-100/061-Rotate List/leetcode061.rb
# Definition for singly-linked list.
# class ListNode
#   attr_accessor :val, :next
#   def initialize(val)
#     @val = val
#     @next = nil
#   end
# end

# @param {ListNode} head
# @param {Integer} k
# @return {ListNode}
def rotate_right(head, k)
  cnt, len, cur = 0, 0, head
  while (cur != nil) do
    if (cur.next == nil) then
      len, cnt = cnt + 1, 0
      cur.next, cur = head, head
      k %= len
    end
    if (cnt == len - k - 1) then
      ret, cur.next = cur.next, nil
      return ret
    end
    cnt, cur = cnt + 1, cur.next
  end
  return head
end